HDC302X - Add driver for this sensor

StijnCoussement opened the pull request:
Add a driver for the TI HDC302X temperature and humidity sensor.
datasheet can be found on
https://cdn-learn.adafruit.com/assets/assets/000/131/097/original/hdc3021-q1.pdf?1720448563

pdgendt commented on drivers/sensor/ti/ti_hdc302x/ti_hdc302x.c:
```diff
@@ -0,0 +1,976 @@
+	return 0;
+}
+
+static DEVICE_API(sensor, ti_hdc302x_api_funcs) = {
+    .sample_fetch = ti_hdc302x_sample_fetch,
+    .channel_get = ti_hdc302x_channel_get,
+    .attr_set = ti_hdc302x_attr_set,
+    .attr_get = ti_hdc302x_attr_get,
+};
```
Use tabs

```suggestion
static DEVICE_API(sensor, ti_hdc302x_api_funcs) = {
	.sample_fetch = ti_hdc302x_sample_fetch,
	.channel_get = ti_hdc302x_channel_get,
	.attr_set = ti_hdc302x_attr_set,
	.attr_get = ti_hdc302x_attr_get,
};
```

pdgendt commented on drivers/sensor/ti/ti_hdc302x/ti_hdc302x.c:
```diff
@@ -0,0 +1,976 @@
+	}
+
+	/* Configure interrupt GPIO if available */
+	if (config->gpio_int.port) {
```
```suggestion
	if (config->gpio_int.port != NULL) {
```

pdgendt commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -456,6 +456,12 @@ test_i2c_hdc2080: hdc2080@42 {
+test_i2c_hdc3022: hdc3022@44 {
+	compatible = "ti,hdc3022";
+	reg = <0x44>;
+    int1-gpios = <&test_gpio 0 0>;
```
Your binding has `int-gpios` as a property. Also indent using tabs.

```suggestion
	int-gpios = <&test_gpio 0 0>;
```

pdgendt commented on include/zephyr/drivers/sensor/ti_hdc302x.h:
```diff
@@ -0,0 +1,113 @@
+	HDC302X_SENSOR_MEAS_INTERVAL_MAX
+};
+
+struct ti_hdc302x_config {
+	struct i2c_dt_spec bus;
+	struct gpio_dt_spec gpio_int;
+};
+
+struct ti_hdc302x_data {
+	struct gpio_callback cb_int;
+
+	/* Semaphore will be used to signal when an interrupt occurs */
+	struct k_sem sem_int;
+
+	uint16_t t_sample;
+	uint16_t rh_sample;
+	uint16_t t_alert;
+	uint16_t rh_alert;
+	uint8_t t_offset;
+	uint8_t rh_offset;
+	enum sensor_power_mode_hdc302x power_mode;
+	enum sensor_measurement_interval_hdc302x interval;
+	uint8_t selected_mode[2];
+};
```
These should not be in a public header, move to the .c file.

pdgendt commented on drivers/sensor/ti/ti_hdc302x/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# TI_HDC302X temperature and humidity sensor configuration options
+
+# Copyright (c) 2025 Psiconrol N.V.
```
```suggestion
# Copyright (c) 2025 Psicontrol N.V.
```

pdgendt commented on drivers/sensor/ti/ti_hdc302x/ti_hdc302x.c:
```diff
@@ -0,0 +1,1011 @@
+/*
+ * Copyright (c) 2025 Psiconrol N.V.
```
```suggestion
 * Copyright (c) 2025 Psicontrol N.V.
```

pdgendt reviewed (commented):
;)

StijnCoussement commented:
I'm on leave after today so it is not urgent. I will pick this up on Monday august 18th. Thanks for the assistance so far.

MaureenHelm reviewed (changes requested):
Thanks for the PR

StijnCoussement commented on drivers/sensor/ti/ti_hdc302x/ti_hdc302x.c:
```diff
@@ -0,0 +1,1017 @@
+struct ti_hdc302x_data {
+	struct gpio_callback cb_int;
+	sensor_trigger_handler_t th_handler;
+	struct sensor_trigger th_trigger;
```
changed it, indeed better this way. changed all other used locations as well...

StijnCoussement commented on dts/bindings/sensor/ti,hdc302x.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2025, Psicontrol N.V.
+# SPDX-License-Identifier: Apache-2.0
+
+description: Texas Instruments HDC302X Temperature and Humidity Sensor
+
+include: [sensor-device.yaml, i2c-device.yaml]
```
hmm seems i forgot to commit that file, sorry.

StijnCoussement commented on drivers/sensor/ti/ti_hdc302x/ti_hdc302x.c:
```diff
@@ -0,0 +1,1013 @@
+		/* Calculate: (offset * 65535 + scale * raw_val) / 65535 */
+		/* Use 64-bit arithmetic to prevent overflow */
+		numerator =
+			(int64_t)params->offset * 65535LL + (int64_t)params->scale * (*raw_value);
```
Yes you are correct, it is not needed to use the LL as it is covered by the explicit int64_t casts. I will update it as well.

StijnCoussement committed:
drivers: sensor: ti: Add TI temp sensor HDC302x

The HDC302X sensor driver is added,
you can use this driver to read temperature and humidity.
Also set an offset, upper and lower limits to get warned
when temperature or humidity get out of band.
The sensor is build for ultra low power applications.

Signed-off-by: COUSSEMENT Stijn <stijn.coussement@psicontrol.com>

pdgendt reviewed (dismissed):
Might need a release note entry

jhedberg commented:
@StijnCoussement this needs a rebase due to a merge conflict

StijnCoussement committed:
drivers: sensor: ti: Add TI temp sensor HDC302x to tests

The HDC302x sensor is added to the i2c tests

Signed-off-by: COUSSEMENT Stijn <stijn.coussement@psicontrol.com>
